### PR TITLE
[SPARK-43782][CORE] Support log level configuration with static Spark conf

### PR DESCRIPTION
**Pull request by jwang0306**
### What changes were proposed in this pull request?

This PR proposes to add a static Spark conf to configure an override to the Log4j logging level. It’s a config version of the `SparkContext.setLogLevel()` semantics, when set it’ll trigger a log level override at the beginning of `SparkContext` startup.

### Why are the changes needed?

This is particularly helpful in the Spark Connect scenario where there’s no way for the client to call `setLogLevel` because `SparkContext` is not yet supported in its API, and when it connects to a platform where it can't change the Log4j config file.

### Does this PR introduce _any_ user-facing change?

Yes, it adds a new public configuration `"spark.log.level"`, and by default it's unset which means there's no override.

### How was this patch tested?

A new unit test in `SparkContextSuite`.

### CODE REVIEW

**Comment by jwang0306** [comment]
cc @grundprinzip @cloud-fan, please take a look.

**Review by grundprinzip — APPROVED** [review]

**Comment by LuciferYang** [comment]
@jwang0306 Sorry I may miss the point, can you give an example to illustrate how we use this configuration in the Spark Connect scenario?

**Comment by grundprinzip** [comment]
It's not meant to be used by spark Connect but simply a convenience option that folks can provide on cluster startup without manipulating the whole log4j.conf which can be quite cumbersome.

**Comment by LuciferYang** [comment]
But the PR description mentions `This is particularly helpful in the Spark Connect scenario` ...

**Comment by grundprinzip** [comment]
Correct, with Spark Connect does not have access to the spark context. For users that want to transition this config it becomes bow very easy. Before they would use set log level as it provides the easiest way to increase or decrease the level. With this option the user has the same surface but using a spark conf.

**Comment by jwang0306** [comment]
@LuciferYang thanks for the question, given the context provided by @grundprinzip, do you have any suggestions on how to update the PR description to better help understand?

> For example launching a cluster with the following command: `./bin/pyspark --conf spark.log.level ERROR`
> In this setup, the Spark Connect client doesn't have access to `SparkContext`, so we can't configure the log level using `sc.setLogLevel()`. But now with this --conf, the launcher is able to set it.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc @rednaxelafx @dongjoon-hyun FYI

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for making a PR, @jwang0306 

1. Although I understand the goal, `OVERRIDE_LOG_LEVEL` sounds misleading to me because this is not the final value. This configuration can be ignored when the user has `.setLogLevel()` invocation in the user code.  

3. If we want to add this, please add documentation like `spark.log.callerContext`.
https://github.com/apache/spark/blob/9d1e67cc200b9315dccdc2f081549dbfe5d1ecd9/docs/configuration.md?plain=1#L406

3. At least for the shell environments, this configuration causes a behavior conflict. Of course, I understand that was this PR's original goal. However, can we mitigate the existing message like `Setting default log level to "WARN".`? As you see in the following, Spark will make the users confused because it says `Setting default log level to "WARN".`, then shows `DEBUG` level messages without proper information. Maybe, this PR may add some proper warning message at line 399 (SparkContext.scala).
```
$ bin/spark-submit -c spark.log.level=TRACE
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
23/05/25 18:37:20 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/05/25 18:37:20 INFO ResourceUtils: ==============================================================
23/05/25 18:37:20 INFO ResourceUtils: No custom resources configured for spark.driver.
23/05/25 18:37:20 INFO ResourceUtils: ==============================================================
23/05/25 18:37:20 INFO SparkContext: Submitted application: Spark shell
23/05/25 18:37:20 INFO ResourceProfile: Default ResourceProfile created, executor resources: Map(cores -> name: cores, amount: 1, script: , vendor: , memory -> name: memory, amount: 1024, script: , vendor: , offHeap -> name: offHeap, amount: 0, script: , vendor: ), task resources: Map(cpus -> name: cpus, amount: 1.0)
23/05/25 18:37:20 INFO ResourceProfile: Limiting resource is cpu
23/05/25 18:37:20 INFO ResourceProfileManager: Added ResourceProfile id: 0
23/05/25 18:37:20 INFO SecurityManager: Changing view acls to: dongjoon
23/05/25 18:37:20 INFO SecurityManager: Changing modify acls to: dongjoon
23/05/25 18:37:20 INFO SecurityManager: Changing view acls groups to:
23/05/25 18:37:20 INFO SecurityManager: Changing modify acls groups to:
23/05/25 18:37:20 INFO SecurityManager: SecurityManager: authentication disabled; ui acls disabled; users with view permissions: dongjoon; groups with view permissions: EMPTY; users with modify permissions: dongjoon; groups with modify permissions: EMPTY
23/05/25 18:37:21 DEBUG InternalLoggerFactory: Using SLF4J as the default logging framework
23/05/25 18:37:21 DEBUG InternalThreadLocalMap: -Dio.netty.threadLocalMap.stringBuilder.initialSize: 1024
23/05/25 18:37:21 DEBUG InternalThreadLocalMap: -Dio.netty.threadLocalMap.stringBuilder.maxSize: 4096
23/05/25 18:37:21 DEBUG MultithreadEventLoopGroup: -Dio.netty.eventLoopThreads: 20
23/05/25 18:37:21 DEBUG GlobalEventExecutor: -Dio.netty.globalEventExecutor.quietPeriodSeconds: 1
23/05/25 18:37:21 DEBUG PlatformDependent0: -Dio.netty.noUnsafe: false
23/05/25 18:37:21 DEBUG PlatformDependent0: Java version: 8
23/05/25 18:37:21 DEBUG PlatformDependent0: sun.misc.Unsafe.theUnsafe: available
23/05/25 18:37:21 DEBUG PlatformDependent0: sun.misc.Unsafe.copyMemory: available
23/05/25 18:37:21 DEBUG PlatformDependent0: sun.misc.Unsafe.storeFence: available
23/05/25 18:37:21 DEBUG PlatformDependent0: java.nio.Buffer.address: available
23/05/25 18:37:21 DEBUG PlatformDependent0: direct buffer constructor: available
23/05/25 18:37:21 DEBUG PlatformDependent0: java.nio.Bits.unaligned: available, true
```

**Comment by jwang0306** [comment]
@dongjoon-hyun thanks for the review,
1. This config statically overrides the log level, but that doesn't  prevent users from dynamically setting new overrides with `setLogLevel`. If you have a strong preference, we can change it to a name without the override word in it, what would you suggest?
2. Sounds good, but let's decide whether to add it or not after (3).
3. That's a very good point. To solve this inconsistency, I think we have multiple alternatives:
    - That line of log happens when `isInterpreter` is true and `silent` = false, however, these are not fields in the Logging trait. So we can't read them in `SparkContext` to warn users about the override under the exact same condition. Maybe we can made the `isInterpreter` and `silent` fields in the Logging trait, or add a new filed to tell whether or not the REPL warning is printed. If we don't care about the original condition (i.e., we don't care about the `silent` mode), then we can print the warning log unconditionally like you suggested.
    - Alternatively, the Logging trait itself might be a better place to implement the override, but `SparkConf` is not available when the trait is initializing. So if we choose to configure the log level override there, we'll have to change the conf to use something else that's already available at that point, like environment variable or Java System Properties. WDYT? cc. @grundprinzip @HyukjinKwon @cloud-fan @LuciferYang

**Comment by dongjoon-hyun** [comment]
Give the config name, `spark.log.level`, simply `LOG_LEVEL` or `SPARK_LOG_LEVEL`?

> This config statically overrides the log level, but that doesn't prevent users from dynamically setting new overrides with setLogLevel. If you have a strong preference, we can change it to a name without the override word in it, what would you suggest?

**Comment by jwang0306** [comment]
@dongjoon-hyun thanks for the suggestion, the variable is now renamed.

For (3), I think I slightly prefer adding a new field in `Logging` to check whether the warning message is printed, so we could print the warning message under the same condition. Does this sounds good to you?

```diff
diff --git a/core/src/main/scala/org/apache/spark/SparkContext.scala b/core/src/main/scala/org/apache/spark/SparkContext.scala
index e142d79523..34fdd4fc32 100644
--- a/core/src/main/scala/org/apache/spark/SparkContext.scala
+++ b/core/src/main/scala/org/apache/spark/SparkContext.scala
@@ -396,7 +396,12 @@ class SparkContext(config: SparkConf) extends Logging {
 
   try {
     _conf = config.clone()
-    _conf.get(SPARK_LOG_LEVEL).foreach(setLogLevel(_))
+    _conf.get(SPARK_LOG_LEVEL).foreach { level =>
+      if (Logging.setLogLevelPrinted) {
+        System.err.printf("Setting Spark log level to \"%s\".\n", level)
+      }
+      setLogLevel(level)
+    }
     _conf.validateSettings()
     _conf.set("spark.app.startTime", startTime.toString)
 
diff --git a/core/src/main/scala/org/apache/spark/internal/Logging.scala b/core/src/main/scala/org/apache/spark/internal/Logging.scala
index 614103dee7..fa00783a9d 100644
--- a/core/src/main/scala/org/apache/spark/internal/Logging.scala
+++ b/core/src/main/scala/org/apache/spark/internal/Logging.scala
@@ -139,6 +139,7 @@ trait Logging {
             context.setConfigLocation(url.toURI)
             if (!silent) {
               System.err.println(s"Using Spark's default log4j profile: $defaultLogProps")
+              Logging.setLogLevelPrinted = true
             }
           case None =>
             System.err.println(s"Spark was unable to load $defaultLogProps")
@@ -164,6 +165,7 @@ trait Logging {
             System.err.printf("Setting default log level to \"%s\".\n", replLevel)
             System.err.println("To adjust logging level use sc.setLogLevel(newLevel). " +
               "For SparkR, use setLogLevel(newLevel).")
+            Logging.setLogLevelPrinted = true
           }
           Logging.sparkShellThresholdLevel = replLevel
           rootLogger.getAppenders().asScala.foreach {
@@ -189,6 +191,7 @@ private[spark] object Logging {
   @volatile private var defaultRootLevel: Level = null
   @volatile private var defaultSparkLog4jConfig = false
   @volatile private[spark] var sparkShellThresholdLevel: Level = null
+  @volatile var setLogLevelPrinted: Boolean = false
 
   val initLock = new Object()
   try {
```

**Comment by dongjoon-hyun** [comment]
So, does the proposal still have three lines of logs like the following when `spark.log.level=TRACE`?
```
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
...
Setting Spark log level to "TRACE".
```

Is the main difficult that `Logging` class doesn't have `SparkConf`?

**Comment by jwang0306** [comment]
>  does the proposal still have three lines of logs like the following when spark.log.level=TRACE in spark-shell/pyspark/sparkr shells?

Yes.

> Is the main difficult that Logging class doesn't have SparkConf?

Yes, the `Logging` trait can be mixed in to multiple places, and unfortunately it’s not feasible to take in `SparkConf` in that trait.

**Comment by dongjoon-hyun** [comment]
Got it. Given the limitation, +1 for the proposed approach. Thank you, @jwang0306 .

File: docs/configuration.md [hunk 1]
```diff
@@ -412,6 +412,15 @@ of the most common options to set are:
...
 </tr>
+<tr>
+  <td><code>spark.log.level</code></td>
+  <td>(none)</td>
+  <td>
+    When set, overrides any user-defined log settings as if calling
+    <code>SparkContext.setLogLevel()</code> at Spark startup. Valid log levels include: "ALL", "DEBUG", "ERROR", "FATAL", "INFO", "OFF", "TRACE", "WARN".
+  </td>
+  <td>3.5.0</td>
+</tr>
```
Comment (dongjoon-hyun) [inline docs/configuration.md:423]: Thanks!

**Comment by jwang0306** [comment]
@dongjoon-hyun both (2) and (3) are now updated, please take another look.

File: core/src/main/scala/org/apache/spark/internal/Logging.scala [hunk 2]
```diff
@@ -189,6 +191,7 @@ private[spark] object Logging {
   @volatile private var defaultRootLevel: Level = null
   @volatile private var defaultSparkLog4jConfig = false
   @volatile private[spark] var sparkShellThresholdLevel: Level = null
+  @volatile var setLogLevelPrinted: Boolean = false
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/Logging.scala:194]: `private[spark]`?

**Review by dongjoon-hyun — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 3]
```diff
@@ -2679,7 +2685,7 @@ class SparkContext(config: SparkConf) extends Logging {
  * various Spark features.
  */
 object SparkContext extends Logging {
-  private val VALID_LOG_LEVELS =
+  val VALID_LOG_LEVELS =
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:2688]: Given the use case, the following will be sufficient.
```scala
-  val VALID_LOG_LEVELS =
+  private[spark] val VALID_LOG_LEVELS =
```

Comment (jwang0306) [reply core/src/main/scala/org/apache/spark/SparkContext.scala]: Thanks! It's now updated.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by jwang0306** [comment]
Gentle ping @dongjoon-hyun, I further added `private[spark]` to the newly added field in `Logging`. Would you mind taking another pass?

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @jwang0306 , @HyukjinKwon , @grundprinzip .
Merged to master for Apache Spark 3.5.0.


And, welcome to the Apache Spark community, @jwang0306 . I added you to the Apache Spark contributor group and assigned SPARK-43782 to you.

**Comment by jwang0306** [comment]
> And, welcome to the Apache Spark community, @jwang0306 . I added you to the Apache Spark contributor group and assigned [SPARK-43782](https://issues.apache.org/jira/browse/SPARK-43782) to you.

Many thanks!

**Comment by mehulbatra-d11** [comment]
@dongjoon-hyun Currently we are facing trouble while syncing events logs to s3 using Spark Connect, Is it expected that with Spark 3.5 it's currently not possible to move events logs to s3? or is there any property that we might be missing?

**Pull request closed without merging**